compose-box: Insert quoted content at the cursor position.

garg3133 opened the pull request:
Right now, on clicking `quote and reply` on any message, the quoted
message is always inserted at the top of compose-box irrespective of
the current cursor position. Also, after insertion of the quoted
message, the cursor is shifted at the end of the compose box.

This commit changes this behaviour to insertion of quoted message
at the current cursor position with a newline at the end of quote
and moving the cursor position to that newline after insertion.
A newline is added at the beginning of quoted message only if there
was some content already present in compose box before the previous
cursor position.

Tested on Google Chrome and Firefox browsers on Ubuntu dev environment.

Fixes: #16836


![quote-reply (1)](https://user-images.githubusercontent.com/39924567/103410773-d7a2a100-4b92-11eb-8593-2f9086ad6064.gif)

garg3133 commented:
I'll fix the failing tests once the approach to resolve this issue is finalised.

You may follow the conversation here: https://chat.zulip.org/#narrow/stream/3-backend/topic/Quote.20.26.20Reply

timabbott commented on static/js/compose_actions.js:
```diff
@@ -445,6 +443,8 @@ exports.quote_and_reply = function (opts) {
         compose_ui.replace_syntax("[Quoting…]", content, textarea);
         autosize.update($("#compose-textarea"));
+        // Update textarea caret to point to the new line after quoted message.
+        textarea.caret(prev_caret + content.length + 1);
```
I'm suspicious of adding `1` here -- isn't that only correct when `prev_caret` was 0?

garg3133 replied:
No, +1 is for the `\n` that is being added after the content in this line:
```
 compose_ui.insert_syntax_and_focus("[Quoting…]\n", textarea);
```
as the `replace_content` function only replaces `[Quoting...]` and so the `\n` is not counted in `content.length`.

timabbott replied:
Wait, so is this a separate pre-existing bug? If so, can you fix it in a separate commit with a commit message explaining the issue? It's really helpful to not combine changes for making reviewers able to efficiently understand why a set of changes is correct.

garg3133 replied:
Sure, will do.

garg3133 replied:
Reflecting on this, I don't think it is a bug. The purpose of adding a newline after `[Quoting…]` is that while quoting a message, the `[Quoting…]` text is immediately added to the compose-box while the insertion of the actual quoted message may take time (depending on internet speed) so it provides a feedback to the user that the quote is in process of being added. 

So, not adding a trailing newline here would mix up this feedback with the rest of the message.

timabbott commented on static/js/compose_actions.js:
```diff
@@ -416,21 +416,19 @@ exports.quote_and_reply = function (opts) {
         exports.respond_to_message(opts);
     }
 
+    const prev_caret = textarea.caret();
```
Should we create this variable above the last block, so that we can use it?  Or is the intent that the value be after the newline that is potentially added above?

garg3133 replied:
Yes, the intent here is to take the value after the potential newline from the previous block is added. If we put this before the previous block, then we'll have to modify `prev_caret` if the newline is added.

timabbott commented on static/js/compose_actions.js:
```diff
@@ -416,21 +416,19 @@ exports.quote_and_reply = function (opts) {
+            // Insert a newline before quoted message if there is
+            // already some content in the compose box and quoted
+            // message is not being inserted at the beginning.
+            textarea.caret("\n");
```
Reflecting on this, shouldn't we just have this add a `\n` to the start of the content we're going to add?  

I think that might end up being simpler.

garg3133 replied:
Yeah, it will be much simpler that way. The only issue is that it will always add a newline before the quote that way even if the compose box was initially empty or the quoted message is being added at the initial position. Although any initial newlines are removed from the message once the message is sent but it does appear while writing the message.

If that's not a problem, we can have the newline to be always added before the content.

timabbott replied:
Well, I was imagining we'd do this conditionally still, just the mechanism would be adding a leading newline rather than doing this `textarea.caret("\n")` here, that generates an off-by-one thing we need to correct elsewhere.

garg3133 replied:
Issues with adding a leading newline at the start of `content` instead of here:
* The feedback given the users by adding `[Quoting…]` will get mixed up with the content already present in compose-box (while it should be visible separately).
* While inserting `[Quoting…]` in the compose-box, a leading space may/may not be added to `[Quoting…]` (which ensures that the text being inserted starts with a new word) so it would be difficult to predict where to set the final caret position in the compose-box.

Also, I don't understand the problem with adding a newline here. Can you please explain it to me?

garg3133 commented:
**How it works now**

Right now, on clicking the `Quote & Reply` option on any message, the following steps takes place:
* If the compose-box was either closed or was open but with no content in it, a new compose-box opens with initial caret position set to 0.
* If a compose-box was already open with some content in it, the caret position is set to 0 (so as to put the quoted text at the start of the message).
* Then, `[Quoting…]` is inserted at the caret position, with a **trailing newline** so as to separate it from rest of the message.
https://github.com/zulip/zulip/blob/bc6c83c12e39b898f5b930f85f8622d783f7d6e5/static/js/compose_actions.js#L434
* After that, `replace_content` function is run which produces the quoted text as we see it and stores it in `content`.
* The `[Quoting…]` text we added earlier is replaced with `content` here:
https://github.com/zulip/zulip/blob/bc6c83c12e39b898f5b930f85f8622d783f7d6e5/static/js/compose_actions.js#L446
* The above line also shifts the caret position to the end of compose-box once the `[Quoting…]` is replaced with the actual quote.

**How it will work after this PR**

On clicking the `Quote & Reply` option on any message, the following steps will takes place:
* If the compose-box was either closed or was open but with no content in it, a new compose-box opens with initial caret position set to 0.
* If a compose-box was already open with some content in it:
  * If the current caret position is 0 (user wants to add the quote at the start of content), nothing happens and the caret position remains 0.
  * If the current caret position is not 0 (user wants to add the quote somewhere in between the content or at the end), a newline is inserted at the caret position (so as to start the quote from a separate line) and the caret position changes to the position next to the added newline.
* The current caret position (after possible addition of newline in last point) is stored in `prev_caret`.
* `[Quoting…]` is inserted at the current caret position, with a **trailing newline** so as to separate it from rest of the trailing content.
https://github.com/zulip/zulip/blob/bc6c83c12e39b898f5b930f85f8622d783f7d6e5/static/js/compose_actions.js#L434
* After that, `replace_content` function is run which produces the quoted text as we see it and stores it in `content`.
* The `[Quoting…]` text we added earlier is replaced with `content` here:
https://github.com/zulip/zulip/blob/bc6c83c12e39b898f5b930f85f8622d783f7d6e5/static/js/compose_actions.js#L446
* The above line also shifts the caret position to the end of compose-box once the `[Quoting…]` is replaced with the actual quote. This is corrected by re-setting the caret position to (`prev_caret` + length of `content` which replaces `[Quoting…]` + `1` for newline added with `[Quoting…]`.

timabbott commented:
OK, I think this makes sense; it's just really subtle, and as a result hard to think about which potentially more readable refactors might be correct.  I think what I'd really like to see happen for this PR is our adding a node test suite (maybe to the `compose_ui.js` suite?) that lets us verify that the cursor position ends up where we want it to be.  Ideally something that lets us write a test with pseudocode like this:
```
set_compose_content_with_caret("hello |there"); // | is used to encode/display position of focus before change
compose_actions.quote_and_reply(...)
assert(get_compose_content_with_caret(), "...");
```

and see that the caret ended up in the right place.  Doing this will make it way way easier for us to iterate on this logic, which is already really subtle in a way that makes me worry merging a PR like this will create some other random bug.  Ideally all of the nice examples you put together above could just be test cases in this new suite.

@YashRE42 would you be up for helping @garg3133 get such a test suite going to help unblock this PR?

YashRE42 commented:
This seems to be sufficient for the helpers:

```javascript
let compose_textbox;
function set_compose_content_with_caret(content) {
    const caret_position = content.indexOf("|");
    content = content.slice(0, caret_position) + content.slice(caret_position + 1); // remove the "|"
    compose_textbox = make_textbox(content);
    compose_textbox.caret(caret_position);
    compose_textbox.trigger("focus");
}
function get_compose_content_with_caret() {
    const caret_position = compose_textbox.caret();
    const content = compose_textbox.val().slice(0, caret_position) + "|" + compose_textbox.val().slice(caret_position); // insert the "|"
    return content;
}

set_compose_content_with_caret("hello |there"); // "|" is used to encode/display position of focus before change
assert(get_compose_content_with_caret(), "hello |there")
```

YashRE42 commented:
Spending a bit more time on this, I came up with https://github.com/zulip/zulip/pull/17864, but with that change, the node tests are expecting for the caret to be at the start of the compose box text, which is wrong but I haven't yet been able to figure out why there's a difference.

timabbott commented:
Now that #19036 was merged, you can rebase this on top of that.

garg3133 commented:
Rebased on top of #19036. Tested it on my Ubuntu development server and everything works as expected.

The only thing that remains in this PR is the updation of tests to cover other cases as well, like if the compose-box was closed before quoting a message or it only contained text with like ` \n\n  \n` (only space and newline characters), the compose-box should re-open with the caret positioned at 0.

Working on this now.

garg3133 commented:
Updated the tests to cover all possible cases.

timabbott commented:
Merged, thanks @garg3133!
